client/asset/dcr/dcr.go

```diff
@@ -0,0 +1,1181 @@
+		if err != nil {
+			return fmt.Errorf("error decoding redeem script for %s, script = %s: %v", unspent.rpc.TxID, unspent.rpc.RedeemScript, err)
+		}
+		op := newOutput(dcr.node, txHash, unspent.rpc.Vout, v, unspent.rpc.Tree, redeemScript)
+		coins = append(coins, op)
+		spents = append(spents, wire.NewOutPoint(txHash, op.vout, unspent.rpc.Tree))
+		size += unspent.input.Size()
+		sum += v
+		return nil
+	}
+
+out:
+	for {
+		// On each loop, find the smallest UTXO that is enough for the value. If
+		// no UTXO is large enough, add the largest and continue.
+		var txout *compositeUTXO
+		for _, txout = range utxos {
+			if isEnoughWith(txout) {
+				err = addUTXO(txout)
+				if err != nil {
+					return nil, err
+				}
+				break out
+			}
+		}
+		// Append the last output, which is the largest.
+		err = addUTXO(txout)
+		if err != nil {
+			return nil, err
+		}
+		// Pop the utxo from the unspents
+		utxos = utxos[:len(utxos)-1]
+		// If there are none left, we don't have enough.
+		if len(utxos) == 0 {
+			return nil, fmt.Errorf("not enough to cover requested funds + fees = %d",
+				dcr.reqFunds(value, size))
+		}
+	}
+
+	err = dcr.node.LockUnspent(false, spents)
+	if err != nil {
+		return nil, err
+	}
+	return coins, nil
+}
+
+// ReturnCoins unlocks coins. This would be necessary in the case of a
+// canceled order.
+func (dcr *ExchangeWallet) ReturnCoins(unspents asset.Coins) error {
+	if len(unspents) == 0 {
+		return fmt.Errorf("cannot return zero coins")
+	}
+	ops := make([]*wire.OutPoint, 0, len(unspents))
+
+	for _, unspent := range unspents {
+		op, err := dcr.convertCoin(unspent)
+		if err != nil {
+			return fmt.Errorf("error converting coin: %v", err)
+		}
+		ops = append(ops, wire.NewOutPoint(&op.txHash, op.vout, op.tree))
+	}
+	return dcr.node.LockUnspent(true, ops)
+}
+
+// Swap sends the swaps in a single transaction. The Receipts returned can be
+// used to refund a failed transaction.
+func (dcr *ExchangeWallet) Swap(swaps []*asset.Swap) ([]asset.Receipt, error) {
+	var contracts [][]byte
+	var totalOut, totalIn uint64
+	// Start with an empty MsgTx.
+	baseTx := wire.NewMsgTx()
+	// Add the funding utxos.
+	for _, swap := range swaps {
+		for _, coin := range swap.Inputs {
+			output, err := dcr.convertCoin(coin)
+			if err != nil {
+				return nil, fmt.Errorf("error converting coin: %v", err)
+			}
+			if output.value == 0 {
+				return nil, fmt.Errorf("zero-valued output detected for %s:%d", output.txHash, output.vout)
+			}
+			totalIn += output.value
+			prevOut := wire.NewOutPoint(&output.txHash, output.vout, wire.TxTreeRegular)
+			txIn := wire.NewTxIn(prevOut, int64(output.value), []byte{})
+			baseTx.AddTxIn(txIn)
+		}
+	}
+	// Add the contract outputs.
+	for _, swap := range swaps {
+		contract := swap.Contract
+		totalOut += contract.Value
+		// revokeAddr is the address that will receive the refund if the contract is
+		// abandoned.
+		revokeAddr, err := dcr.node.GetNewAddress(dcr.acct, chainParams)
```

Comment:
As was pointed out at https://github.com/decred/dcrdex/pull/72#discussion_r350363610, the liberal use of new addresses will need some thinking here.